File: lectures/5/Mandelbrot.py
import time
import numpy as np

def mandelbrot(c, max_iter):
    z = 0
    for i in range(max_iter):
        z = z * z + c
        if abs(z) > 2:
            return i
    return max_iter

def mandelbrot_set(width, height, max_iter):
    real_min, real_max = -2.0, 1.0
    imag_min, imag_max = -1.5, 1.5
    real_vals = np.linspace(real_min, real_max, width)
    imag_vals = np.linspace(imag_min, imag_max, height)
    
    result = np.zeros((height, width))
    for i in range(height):
        for j in range(width):
            result[i, j] = mandelbrot(complex(real_vals[j], imag_vals[i]), max_iter)
    return result

if __name__ == "__main__":
    width, height, max_iter = 2000, 2000, 1000
    start_time = time.time()
    mandelbrot_set(width, height, max_iter)
    end_time = time.time()
    print(f"Execution Time: {end_time - start_time:.2f} seconds")

File: projects/3/xavier_baines/lectures/5/Mandelbrot.java
public class Mandelbrot {
    public static int mandelbrot(double real, double imag, int maxIter) {
        double zReal = 0, zImag = 0;
        int iter;
        for (iter = 0; iter < maxIter; iter++) {
            double zReal2 = zReal * zReal - zImag * zImag + real;
            double zImag2 = 2 * zReal * zImag + imag;
            zReal = zReal2;
            zImag = zImag2;
            if (zReal * zReal + zImag * zImag > 4) {
                break;
            }
        }
        return iter;
    }

    public static void mandelbrotSet(int width, int height, int maxIter) {
        double realMin = -2.0, realMax = 1.0;
        double imagMin = -1.5, imagMax = 1.5;
        double realStep = (realMax - realMin) / width;
        double imagStep = (imagMax - imagMin) / height;

        int[][] result = new int[height][width];
        for (int i = 0; i < height; i++) {
            for (int j = 0; j < width; j++) {
                double real = realMin + j * realStep;
                double imag = imagMin + i * imagStep;
                result[i][j] = mandelbrot(real, imag, maxIter);
            }
        }
    }

    public static void main(String[] args) {
        int width = 2000, height = 2000, maxIter = 1000;
        long startTime = System.currentTimeMillis();
        mandelbrotSet(width, height, maxIter);
        long endTime = System.currentTimeMillis();
        System.out.println("Execution Time: " + (endTime - startTime) / 1000.0 + " seconds");
    }
}
